evm-contracts/src/v0.6/VRFCoordinator.sol

```diff
@@ -110,70 +116,130 @@ contract VRFCoordinator is VRF, VRFRequestIDBase {
    * @notice creates the chainlink request for randomness
    *
    * @param _keyHash ID of the VRF public key against which to generate output
-   * @param _seed Input to the VRF, from which randomness is generated
+   * @param _consumerSeed Input to the VRF, from which randomness is generated
    * @param _feePaid Amount of LINK sent with request. Must exceed fee for key
    * @param _sender Requesting contract; to be called back with VRF output
+   *
+   * @dev _consumerSeed is mixed with key hash, sender address and nonce to
+   * @dev obtain preSeed, which is passed to VRF oracle, which mixes it with the
+   * @dev hash of the block containing this request, to compute the final seed.
+   *
+   * @dev The requestId used to store the request data is constructed from the
+   * @dev preSeed and keyHash.
    */
   function randomnessRequest(
     bytes32 _keyHash,
-    uint256 _seed,
+    uint256 _consumerSeed,
     uint256 _feePaid,
     address _sender
   )
     internal
     sufficientLINK(_feePaid, _keyHash)
   {
     uint256 nonce = nonces[_keyHash][_sender];
-    uint256 seed = makeVRFInputSeed(_keyHash, _seed, _sender, nonce);
-    bytes32 requestId = makeRequestId(_keyHash, seed);
+    uint256 preSeed = makeVRFInputSeed(_keyHash, _consumerSeed, _sender, nonce);
+    bytes32 requestId = makeRequestId(_keyHash, preSeed);
     // Cryptographically guaranteed by seed including an increasing nonce
     assert(callbacks[requestId].callbackContract == address(0));
     callbacks[requestId].callbackContract = _sender;
-    callbacks[requestId].randomnessFee = _feePaid;
-    callbacks[requestId].seed = seed;
-    emit RandomnessRequest(_keyHash, seed, serviceAgreements[_keyHash].jobID,
+    assert(_feePaid < 1e9); // Total LINK fits in uint96
+    callbacks[requestId].randomnessFee = uint96(_feePaid);
+    callbacks[requestId].seedAndBlockNum = keccak256(abi.encodePacked(
+      preSeed, block.number));
+    emit RandomnessRequest(_keyHash, preSeed, serviceAgreements[_keyHash].jobID,
       _sender, _feePaid);
     nonces[_keyHash][_sender] = nonces[_keyHash][_sender].add(1);
   }
 
+  // Offsets into fulfillRandomnessRequest's _proof of various values
+  //
+  // Public key. Skips byte array's length prefix.
+  uint256 public constant PUBLIC_KEY_OFFSET = 0x20;
+  // Seed is 7th word in proof, plus word for length, (6+1)*0x20=0xe0
+  uint256 public constant PRESEED_OFFSET = 0xe0;
+
   /**
    * @notice Called by the chainlink node to fullfil requests
+   *
    * @param _proof the proof of randomness. Actual random output built from this
    *
-   * @dev This is the main entrypoint for chainlink. If you change this, you
-   * @dev should also change the solidityABISstring in solidity_proof.go.
+   * @dev The structure of _proof corresponds to vrf.MarshaledOnChainResponse,
+   * @dev in the node source code. I.e., it is a vrf.MarshaledProof with the seed
```

Comment:
Should there be a `@dev` on each line?

Reply:
I thought that's what you're supposed to do if you want the docstrings to work with natspec, but  it seems to make no difference to the natspec output, and that output is horrible anyway. I will remove these in another PR, perhaps.